### ISSUE TITLE
Editing a contact leads to 403 error: no-uid-conflict

### ISSUE THREAD

**Issue by Johann-Tree** [inverse-inc/sogo-connector.tb31#2]
I'm using Thunderbird 31.3.0 on Ubuntu 14.04.01 with Inverse SOGo Connector 31.0.1 to synchronize my Addressbook with a Thunderbird installation on another machine with the CardDAV-Server by Synology (on their NAS DS213j with current software DSM 5.1-5022). I copied 50 contacts in the Remote Address Book. They were synced successfully. 
Then I added a new contact. A few moments later I saw, that there was a typo. So I edited the contact but now I got the message that the sync process failed. The error console says: 
```
Upload failure uploading card: https://synerde:8443/addressbooks/users/user/addressbook/C67646F4-5580-0001-C07A-1A40B0F0B2F0.vcf.
HTTP Status Code:403
  * card properties:
  0 prop: [xpconnect wrapped nsIProperty];  name: DisplayName; value: Vorname Nachname
  1 prop: [xpconnect wrapped nsIProperty];  name: RecordKey; value: 274
  2 prop: [xpconnect wrapped nsIProperty];  name: PreferMailFormat; value: 0
  3 prop: [xpconnect wrapped nsIProperty];  name: BirthDay; value: 22
  4 prop: [xpconnect wrapped nsIProperty];  name: groupDavKey; value: C67646F4-5580-0001-C07A-1A40B0F0B2F0.vcf
  5 prop: [xpconnect wrapped nsIProperty];  name: HomeAddress; value: Straße 79
  6 prop: [xpconnect wrapped nsIProperty];  name: LowercasePrimaryEmail; value: email@example.org
  7 prop: [xpconnect wrapped nsIProperty];  name: groupDavVersionPrev; value: "ba09a3b9734a26858586861b1210263b"
  8 prop: [xpconnect wrapped nsIProperty];  name: PreferDisplayName; value: 1
  9 prop: [xpconnect wrapped nsIProperty];  name: BirthYear; value: 1999
  10 prop: [xpconnect wrapped nsIProperty];  name: CellularNumber; value: 0123-1234567
  11 prop: [xpconnect wrapped nsIProperty];  name: PhotoType; value: generic
  12 prop: [xpconnect wrapped nsIProperty];  name: Notes; value: Mein: Text
Mehr: Text
  13 prop: [xpconnect wrapped nsIProperty];  name: FirstName; value: Vorname
  14 prop: [xpconnect wrapped nsIProperty];  name: LastName; value: Nachname
  15 prop: [xpconnect wrapped nsIProperty];  name: CardUID; value: C67646F4-5580-0001-C07A-1A40B0F0B2F0.vcf
  16 prop: [xpconnect wrapped nsIProperty];  name: groupDavVersion; value: -1
  17 prop: [xpconnect wrapped nsIProperty];  name: PopularityIndex; value: 0
  18 prop: [xpconnect wrapped nsIProperty];  name: WebPage2; value: http://example.org
  19 prop: [xpconnect wrapped nsIProperty];  name: PrimaryEmail; value: email@example.org
  20 prop: [xpconnect wrapped nsIProperty];  name: HomePhone; value: 01234-12345
  21 prop: [xpconnect wrapped nsIProperty];  name: BirthMonth; value: 12
  22 prop: [xpconnect wrapped nsIProperty];  name: HomeCity; value: Musterstadt
  23 prop: [xpconnect wrapped nsIProperty];  name: DbRowID; value: 46
  24 prop: [xpconnect wrapped nsIProperty];  name: LastModifiedDate; value: 1421588338
  25 prop: [xpconnect wrapped nsIProperty];  name: HomeZipCode; value: 12345
  * done
```
In the access.log of the CardDAV Server I found:
```
::ffff:192.168.1.194 - user [18/Jan/2015:23:22:38 +0200] "PUT /addressbooks/users/user/addressbook/C67646F4-5580-0001-C07A-1A40B0F0B2F0.vcf HTTP/1.1" 403 271 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:31.0) Gecko/20100101 Thunderbird/31.3.0 Lightning/3.3.2" i=1 or=1 t=112.6 cl=438 err=no-uid-conflict
```

What information can I provide  to help solving this problem?
Thanks!

**Comment by extrafu** [comment]
You need Mozilla Lightning.

**Comment by Johann-Tree** [comment]
I'm sorry, I forgot to mention that Lightning is installed, of course.

**Comment by Johann-Tree** [comment]
The error is reproducible: Editing a contact twice within seconds leads to the error as described above. I use the standard configuration with periodic sync every 15 minutes.
What information can I provide to help solving this problem?

**Comment by Johann-Tree** [comment]
In /volume1/@appstore/CardDAVServer/var/log/caldavd/error.log on Synology DS213j with current software DSM 5.2-5565 (CardDAV Server 5.2.0-0027) I now get this error:

```
2015-05-20 23:29:21+0200 [-] [caldav-0]  [-] [twistedcaldav.vcard#debug] vCard data had unfixable problems:
2015-05-20 23:29:21+0200 [-] [caldav-0] 	  [VCARD] Missing or too many required property: N
```

**Comment by sebastiendamaye** [comment]
Hi,
I had the same issue and found from the RFC (https://tools.ietf.org/html/rfc6352) that the error is indicating a conflict with existing UIDs:

    (CARDDAV:no-uid-conflict): The resource submitted in the PUT
    request, or targeted by a COPY or MOVE request, MUST NOT specify a
      vCard UID property value already in use in the targeted address
      book collection or overwrite an existing address object resource
      with one that has a different UID property value.  Servers SHOULD
      report the URL of the resource that is already making use of the
      same UID property value in the DAV:href element.

A workaround that worked for me was to export the contacts as LDIF, delete all contacts, sync with server, re-import all contacts from the LDIF file and re-sync. Hope this will help.

**Comment by Giermann** [comment]
Although I had a different result (duplicated contacts), this problem seems related to the fact that the Connector does not set the groupDavKey on the very first sync.
Can you try to implement my proposed patch: https://github.com/inverse-inc/sogo-connector.tb31/pull/7

After that, please delete all contacts in this Remote Addressbook both in Thunderbird and on the server (if they still exist), copy all contacts again and try to modify one of them?

**Comment by extrafu** [comment]
Is this issue still relevant after your patches were added?

**Comment by Johann-Tree** [comment]
I would like to test but I have no idea how to build the extension. Any hints highly appreciated!

**Comment by extrafu** [comment]
Simply clone the git repo on Linux and run "make".

**Comment by Giermann** [comment]
@Johann:
Or write a short message to giermann <at> funke <dot> de and I'll send you
the XPI.

2015-10-23 22:54 GMT+02:00 extrafu <notifications@github.com>:

**Comment by Johann-Tree** [comment]
Thanks, I have already built the xpi. As soon as I have new information I will post them here.

**Comment by Johann-Tree** [comment]
Sorry for the delay, there were so many other things to do out on the farm.
Today I was able to finally test it. I'm happy to announce that now it's working like expected! It was also not necessary to delete all contacts. Most of the failed contacts were uploaded right after I installed/updated the extension. Only two contacts I had to delete through Thunderbird and create them again (not copying from another address book.)
Thank you so much!
